ui/packages/shared/profile/src/ProfileIcicleGraph/IcicleGraphArrow/TextWithEllipsis.tsx

```diff
@@ -0,0 +1,91 @@
+// Copyright 2022 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {useEffect, useRef, useState} from 'react';
+
+import {USER_PREFERENCES, useUserPreference} from '@parca/hooks';
+
+interface Props {
+  text: string;
+  x: number;
+  y: number;
+  width: number;
+}
+
+function calculateTruncatedText(
+  text: string,
+  textElement: SVGTextElement,
+  maxWidth: number
+): string {
+  // Create a temporary element for measurement
+  const tempElement = textElement.cloneNode() as SVGTextElement;
+  tempElement.textContent = text;
+  textElement.parentElement?.appendChild(tempElement);
+
+  // If the text fits, return it
+  const fullWidth = tempElement.getComputedTextLength();
+  if (fullWidth <= maxWidth) {
+    textElement.parentElement?.removeChild(tempElement);
+    return text;
+  }
+
+  // Binary search to find the maximum text that fits
+  let start = 0;
+  let end = text.length;
+  let result = text;
+
+  while (start < end) {
+    const mid = Math.floor((start + end + 1) / 2);
+    const truncated = text.slice(-mid);
+
+    tempElement.textContent = truncated;
+    const currentWidth = tempElement.getComputedTextLength();
+
+    if (currentWidth <= maxWidth) {
+      result = truncated;
+      start = mid;
+    } else {
+      end = mid - 1;
+    }
+  }
+
+  textElement.parentElement?.removeChild(tempElement);
+  return result;
+}
+
+function TextWithEllipsis({text, x, y, width}: Props): JSX.Element {
+  const textRef = useRef<SVGTextElement>(null);
+  const [displayText, setDisplayText] = useState(text);
+  const [showFunctionNameFromLeft] = useUserPreference<boolean>(
+    USER_PREFERENCES.SHOW_FUNCTION_NAME_FROM_LEFT.key
+  );
+
+  useEffect(() => {
```

Comment:
This `useState` + `useEffect` combo can be replaced with a single `useMemo` that returns the text that can be used below to render.

Technically, `useState` + `useEffect` method has to render the component twice before it renders the expected output while the `useMemo` approach would do it in a single render. Not a biggie, but it improves the readability as well.

Reply:
good catch, i'll make that change.

Reply:
hmm so I made changes based on the feedback above to now use useMemo.

```tsx
  const displayText = useMemo(() => {
    const textElement = textRef.current;
    if (textElement === null) return text;

    return showFunctionNameFromLeft ? text : calculateTruncatedText(text, textElement, width);
  }, [text, width, showFunctionNameFromLeft, textRef]);
```

but with this new change we don't get to see the truncated text when it should be shown.

I'm assuming this is related to the usage of the `textRef` ref and the fact that useMemo is running before the ref is attached to the DOM, therefore the initial render will always have `textRef.current` as null, and the memoization is preventing subsequent updates.

Reply:
cc @manojVivek

Reply:
Ah yes, I missed the ref. Then lets go with the current implementation only.